common/kokkos-sampler/kp_sampler_skip.cpp

```diff
@@ -1,350 +1,420 @@
+                  << ((beginForCallee == NULL) ? "no" : "yes") << "\n";
+        std::cout << "KokkosP: begin-parallel-scan:     "
+                  << ((beginScanCallee == NULL) ? "no" : "yes") << "\n";
+        std::cout << "KokkosP: begin-parallel-reduce:   "
+                  << ((beginReduceCallee == NULL) ? "no" : "yes") << "\n";
+        std::cout << "KokkosP: end-parallel-for:        "
+                  << ((endForCallee == NULL) ? "no" : "yes") << "\n";
+        std::cout << "KokkosP: end-parallel-scan:       "
+                  << ((endScanCallee == NULL) ? "no" : "yes") << "\n";
+        std::cout << "KokkosP: end-parallel-reduce:     "
+                  << ((endReduceCallee == NULL) ? "no" : "yes") << "\n";
+      }
+    }
+  }
+
+  free(envBuffer);
+
+  uniqID = 1;
+
+  if (0 > tool_seed) {
+    srand(time(NULL));
+    if (tool_verbosity > 0) {
+      std::cout << "KokkosP: Seeding random number generator using clock for "
+                   "random sampling.\n";
+    }
+  } else {
+    srand(tool_seed);
+    if (tool_verbosity > 0) {
+      std::cout << "KokkosP: Seeding random number generator using seed "
+                << tool_seed << " for random sampling.\n";
+    }
+  }
+
+  const char* tool_probability = getenv("KOKKOS_TOOLS_SAMPLER_PROB");
+
+  if (NULL != tool_probability) {
+    //  Read sampling probability as a float between 0 and 100, representing
+    //  a percentage that data should be gathered.
+    //  Utility reasons about probability as a double between 0.0 and 1.0.
+    tool_prob_num = atof(tool_probability);
+    if (tool_prob_num > 100.0) {
+      std::cout << "KokkosP: The sampling probability value is set to be "
+                   "greater than 100.0. The probability for the sampler will "
+                   "be set to 100 percent; all of the invocations of a Kokkos "
+                   "kernel will be profiled.\n";
+      tool_prob_num = 100.0;
+    } else if (tool_prob_num < 0.0) {
+      std::cout
+          << "KokkosP: The sampling probability value is set to be a negative "
+             "number. The sampler's probability will be set to 0 percent; none "
+             "of the invocations of a Kokkos kernel will be profiled.\n";
+      tool_prob_num = 0.0;
+    }
+    if (tool_verbosity > 0) {
+      std::cout << "KokkosP: Probability for the sampler set to: "
+                << tool_prob_num << "\n";
+    }
+    kernelSampleSkip = 1;
+    return;
+  }
+
+  const char* tool_sample = getenv("KOKKOS_TOOLS_SAMPLER_SKIP");
+  if (NULL != tool_sample) {
+    tool_prob_num    = 100.0;
+    kernelSampleSkip = atoi(tool_sample) + 1;
+    if (tool_verbosity > 0) {
+      std::cout << "KokkosP: Sampling rate set to: " << tool_sample << "\n";
+    }
+    return;
+  }
+
+  if (tool_verbosity > 0) {
+    std::cout << "KokkosP: Neither the probability nor the skip rate for "
+                 "sampling were set...\n";
+  }
+  tool_prob_num    = 10.0;
+  kernelSampleSkip = 1;
```

Comment:
You hard code both tool_prob_num and kernelSampleSkip ... the user settings are ignored.

Reply:
Put the entire thing into a `if (tool_prob_num == -1.0)`

Reply:
Fixed. 

I think the key suggestion that handles the issue raised is handled by the single one you suggested in the end of this review. 

We might want to catch the case when tool_prob_num is negative but not -1.0 - I think that should happen only if there is some non-deterministic bit flip like event that causes `tool_prob_num` to be negative. 

Note I have now also put comments for each possible user input and how it is handled.